### ISSUE TITLE
Styling Link components in react router with style props has warning about unknown props on a tag

### ISSUE THREAD

**Issue by dmngu9** [styled-components/styled-components#1814]
i used styled(Link) with style props and has this warning
Warning: React does not recognize the xx prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase  instead. If you accidentally passed it from a parent component, remove it from the DOM element.
    in a (created by Link)
    in Link (created by Styled(Link))

Any work around suggestions?:pensive:

**Comment by jameswyse** [comment]
Just ran into this as well with `<NavLink>`, it seems react-router is blindly passing all props to the underlying HTML element, which isn't a good idea for exactly this reason.

I wrote a quick HOC to only pass valid props, it'll do until react-router fixes the problem :)

```javascript
import { NavLink as RouterNavLink } from 'react-router-dom';

const NavLink = props => {
  const passProps = Object.keys(RouterNavLink.propTypes).reduce(
    (passedProps, propName) => {
      if (props.hasOwnProperty(propName)) {
        passedProps[propName] = props[propName];
      }

      return passedProps;
    },
    {}
  );

  return <RouterNavLink {...passProps}>{props.children}</RouterNavLink>;
};
```

**Comment by dmngu9** [comment]
btw, your HOC wont allow pass style prop to Link component though. I wrap the <span> tag around the Link component and has style prop on the <span/> tag

**Comment by probablyup** [comment]
Related to https://github.com/styled-components/styled-components/pull/1682

**Comment by praxxis** [comment]
If you're using TypeScript 2.8+ you can use the new continual types to [create some TypeScript helper types](https://stackoverflow.com/a/50136346) that allow fairly succinct wrapper components. 

For example, here's adding a `currentRoute` prop to a react-router `NavLink`:

```typescript
const AppLink: React.SFC<{currentRoute: boolean} & GetComponentClassProps<typeof NavLink>> = ({currentRoute, children, ...rest}) => {
  currentRoute; // tslint:disable-line:no-unused-expression
  return <NavLink {...rest}>{children}</NavLink >;
};

const StyledAppLink = styled(CurrentAppLink)`
  text-decoration: ${(props) => props.currentRoute ? 'underline' : 'none'};
`;
```

`StyledAppLink` will have type safety covering the props of `NavLink` as well as the `currentRoute` prop.

**Comment by wmertens** [comment]
Should this issue even be open? It's a fairly close duplicate of #1682
